qlib/TextWrap.qm

```diff
@@ -0,0 +1,664 @@
+      drop_whitespace (default: true)
+        Drop leading and trailing whitespace from lines.
+      max_lines (default: None)
+        Truncate wrapped lines.
+      placeholder (default: ' [...]')
+        Append to the last line of truncated text.
+*/
+public class TextWrapper {
+    public {
+        int width = 70;
+        string initial_indent = "";
+        string subsequent_indent = "";
+        bool expand_tabs = True;
+        bool replace_whitespace = True;
+        bool fix_sentence_endings = False;
+        bool break_long_words = True;
+        bool drop_whitespace = True;
+        bool break_on_hyphens = True;
+        int tabsize = 8;
+        int max_lines;
+        string placeholder = ' [...]';
+    }
+
+    private {
+        hash whitespace_trans;
+        string wordsep_re;
+        string wordsep_simple_re;
+        string sentence_end_re;
+    }
+
+    constructor (*hash opts) {
+        if (opts.hasKey('width'))
+            self.width = opts.width;
+        if (opts.hasKey('initial_indent'))
+            self.initial_indent = opts.initial_indent;
+        if (opts.hasKey('subsequent_indent'))
+            self.subsequent_indent = opts.subsequent_indent;
+        if (opts.hasKey('expand_tabs'))
+            self.expand_tabs = opts.expand_tabs;
+        if (opts.hasKey('replace_whitespace'))
+            self.replace_whitespace = opts.replace_whitespace;
+        if (opts.hasKey('fix_sentence_endings'))
+            self.fix_sentence_endings = opts.fix_sentence_endings;
+        if (opts.hasKey('break_long_words'))
+            self.break_long_words = opts.break_long_words;
+        if (opts.hasKey('drop_whitespace'))
+            self.drop_whitespace = opts.drop_whitespace;
+        if (opts.hasKey('break_on_hyphens'))
+            self.break_on_hyphens = opts.break_on_hyphens;
+        if (opts.hasKey('tabsize'))
+            self.tabsize = opts.tabsize;
+        if (opts.hasKey('max_lines'))
+            self.max_lines = opts.max_lines;
+        if (opts.hasKey('placeholder'))
+            self.placeholder = opts.placeholder;
+
+        for (int i=0; i < c_whitespace.length(); i++)
+            self.whitespace_trans.(ord(c_whitespace[i])) = ord(' ');
+
+        # This funky little regex is just the trick for splitting
+        # text up into word-wrappable chunks.  E.g.
+        #   "Hello there -- you goof-ball, use the -b option!"
+        # splits into
+        #   Hello/ /there/ /--/ /you/ /goof-/ball,/ /use/ /the/ /-b/ /option!
+        # (after stripping out empty strings).
+        string word_punct = '[\w!"'"'&.,?]";
+        string letter = '[^\d\W]';
+        string whitespace = sprintf ('[%s]', regex_escape (c_whitespace));
+        string nowhitespace = '[^' + whitespace.substr(1);
+        wordsep_re = xsprintf('
+            ( # any whitespace
+              %(ws)s+
+            | # em-dash between words
+              (?<=%(wp)s) -{2,} (?=\w)
+            | # word, possibly hyphenated
```

Comment:
If these comments are meant as Qore comments, then they won't work because the current implementation of scanner simply takes everything between the two edge `'` symbols...